[c10d] Mixed precision DDP hang fix and fine-grained option for DDP perf

teng-li opened the pull request:
When go to mixed precision fp16 training, DDP randomly hangs.  Initially, I thought this smells like a similar NCCL bug I filed a while ago. It turns out it's not. Again, I am seeing different rank process has different size.  How could this even happen?

It turns out that take_tensors will generate a list of bucketed tensors in an un deterministic order, because, the key to the map is a pointer.  An interesting bug digging and fix.

Now fp16 DDP training should be fully working now.

Also, added another take_tensor fine grained helper that aims to improve the performance of DDP, making it a TODO to replace the DDP take_tensors with that.

Fixed: https://github.com/pytorch/pytorch/issues/12150

teng-li committed:
[c10d] Mixed precision DDP hang fix and fine-grained option for perf improvement

pietern commented on torch/csrc/utils/tensor_flatten.cpp:
```diff
@@ -21,10 +23,11 @@ std::vector<TensorGroup> take_tensors(TensorList tensors, size_t size_limit) {
     }
-    auto & type_group = groups[&type];
+
+    auto & type_group = groups[type.ID()];
```
Not clang-formatted

teng-li replied:
done in the third commit

pietern commented on torch/csrc/utils/tensor_flatten.cpp:
```diff
@@ -39,6 +42,52 @@ std::vector<TensorGroup> take_tensors(TensorList tensors, size_t size_limit) {
+
+    if (currentGroupSize >= sizeLimit) {
+      for (auto& entry : currentGroup){
+        std::cout << "yes" << std::endl;
```
Debug info

teng-li replied:
Done in the second commit

pietern reviewed (commented):
Can you put the finegrained version in a different PR when you use it somewhere?

I don't see how it is different btw. Would be good to add a comment describing it.

Nice find on the `std::unordered_map` problem!

teng-li commented:
@pietern , I think it's OK to keep the other function here. Please see the comments I added for each function in header for differences. I believe theoretically, the function I added can make fp16 DDP faster.

teng-li commented:
@pietern  It's not just unordered_map, it's a problem of using a pointer as the key to that map. No wonder the randomness of hang

pietern commented on torch/csrc/utils/tensor_flatten.h:
```diff
@@ -37,14 +41,35 @@ struct TensorGroup {
+// Helper function that take a list of tensor by the size limit, if the
+// list of tensors are of different tensor types, each type will be taken by
+// the size limit. In other words, the input tensor list will firstlbe bucketed
+// by tensor types, and then each bucket will be further bucketed by size_limit
```
Can you clarify further what _taken by the size limit_ means. I have enough context to know what's going on but can see how a new person reading this doesn't understand what goes on. Can you mention size limit is in bytes, and what an example output is. E.g. `vector{vector{f16}, vector{f16, f16, f16}, vector{f32, f32}}` or something along those lines.

Also some typos.

teng-li replied:
Let me rewrite

pietern commented on torch/csrc/utils/tensor_flatten.h:
```diff
@@ -37,14 +41,35 @@ struct TensorGroup {
+// accumulate enough tensors (even though the tensors can be of different type)
+// until size_limit is hit. Then for each accumulation, it will further bucket
+// the accumulation by different types. In other words, for each size_limit's
+// accumulation, multiple buckets are generated.
```
Multiple buckets as in 1 per type (in the last sentence)? If I understand correctly this will output in chronological order (since you don't care about type when counting up to the size limit), compared to potential complete out of order for the other implementation. If this is the case, I would assume this more fine grained one has a much better shot overlapping indeed and we should consider just replacing it instead of having 2 functions with similar outcome.

teng-li replied:
Well, hold on, the DDP backward bucketing is using the Python implementation of take_tensors, which is different than C++ :D.

Let's keep two for now, I think both can be useful in certain cases.

apaszke replied:
So this basically swaps the order of bucketing compared to the function above? Why would anyone want that?

teng-li replied:
This would be beneficial for BW overlapping, especially for the case of fp16/fp32 mixed precision training.

apaszke commented on torch/csrc/utils/tensor_flatten.cpp:
```diff
@@ -1,44 +1,93 @@
 }
 
+std::vector<TensorGroup> take_tensors_finegrained(
+    TensorList tensors,
```
How does this differ from the function above? There’s a lot of code duplication

teng-li replied:
Eventually, I would like to use the function for BW bucketing strategy, the difference is explained in comments, I will rewrite the comments to make it more clear

apaszke commented on torch/csrc/utils/tensor_flatten.cpp:
```diff
@@ -67,31 +116,37 @@ at::Tensor get_values(const at::Tensor& t) {
+  auto indices =
+      unflatten_dense_tensors(flat_indices, fmap(tensors, &get_indices));
+  auto values =
+      unflatten_dense_tensors(flat_values, fmap(tensors, &get_values));
```
Please avoid reformatting code you’re not touching... it’s not a good style, makes your diff unnecessarily larger, and makes reviewing harder because it adds noise

teng-li commented:
@pietern  Refactored and added comments

teng-li committed:
Added completed tests for both modes and both backends

teng-li commented:
@pietern Completed test added

pietern reviewed (approved):
LGTM

Two additional comments: 1) there's a TODO on using it in DDP (defaults to false now), and 2) the tests can be deduplicated between the two backends. I like the pattern of having a `def _test_dist_broadcast_coalesced(self, args)` and calling it however many times.

pietern commented:
Looks like CircleCI didn't trigger for this PR...

teng-li committed:
Fixed lint errors
